_posts/2024-05-27-A-deep-dive-into-faster-semantic-sparse-retrieval-in-OS-2.12.md

```diff
@@ -0,0 +1,144 @@
+
+### Benchmark results on 1M documents
+
+The following figure shows the experiment results for 1 million documents.
+
+<img src="/assets/media/blog-images/2024-05-27-A-deep-dive-into-faster-semantic-sparse-retrieval-in-OS-2.12/experiments-on-1m.png"/>
+
+### Benchmark results on 8.8M documents
+
+The following figure shows the experiment results for 8.8 million documents.
+
+<img src="/assets/media/blog-images/2024-05-27-A-deep-dive-into-faster-semantic-sparse-retrieval-in-OS-2.12/experiments-on-1m.png"/>
+
+### Results explained
+
+The key takeaways from the experiment are:
+
+1. **For 8.8M documents, search speeds improve significantly for all search methods.** The 8.8M dataset shows significant performance gains from Lucene optimizations. As mentioned previously, query latency includes the time needed for model inference and searching the inverted index. As the number of documents in an index increases, the time needed to search the inverted index grows linearly and becomes the dominant factor in the response time compared to model inference. For the bi-encoder mode, although the model inference time increases due to higher throughput, its overall latency still shows a significant reduction.
+1. **For 8.8M documents, neural sparse search benefits from Lucene upgrades more than BM25.** This is because pruning algorithms like WAND and MAXScore in Lucene struggle to efficiently handle the learned sparse weights used by neural search, compared to the term frequencies that BM25 relies on. Additionally, neural sparse search needs to sort query clauses in order to calculate the minimum competitive score. For older Lucene versions, this step is associated with a greater overhead, hindering performance. Thankfully, [this issue has been addressed](https://github.com/apache/lucene/pull/12490) in newer Lucene versions.
+1. **For 1M documents, we observed a noticeable performance improvement for both BM25 and the doc-only mode of neural sparse search.** However, the gains are less pronounced compared to the 8.8M document set. This is because the overall latency for 1M documents is already relatively low, and some fixed costs, like network I/O and tokenizer inference, take a larger proportion of the total processing time. **For the bi-encoder mode, the overall performance improvement is more limited. In this case, the model inference becomes the bottleneck for query throughput.** Because the model is at the limit of processing requests, some search requests are queued for inference.
```

Comment:
```suggestion
1. **For 1M documents, we observed a noticeable performance improvement for both BM25 and the doc-only mode of neural sparse search.** However, the gains are less pronounced compared to the 8.8M document set. This is because the overall latency for 1M documents is already relatively low, and some fixed costs, like network I/O and tokenizer inference, take a larger proportion of the total processing time. **For the bi-encoder mode, the overall performance improvement is more limited. In this case, the model inference becomes the bottleneck for query throughput.** Because the model is at its request processing limit, some search requests are queued for inference.
```